File: vi/codes/swift/chapter_sorting/heap_sort.swift
/**
 * File: heap_sort.swift
 * Created Time: 2023-05-28
 * Author: nuomi1 (nuomi1@qq.com)
 */

/* Độ dài đống (heap) là n, bắt đầu vun đống nút i, từ trên xuống dưới */
func siftDown(nums: inout [Int], n: Int, i: Int) {
    var i = i
    while true {
        // Nếu nút i là lớn nhất, hoặc chỉ số l, r vượt biên, không cần tiếp tục vun đống, dừng lại
        let l = 2 * i + 1
        let r = 2 * i + 2
        var ma = i
        if l < n, nums[l] > nums[ma] {
            ma = l
        }
        if r < n, nums[r] > nums[ma] {
            ma = r
        }
        // Hoán đổi hai nút
        if ma == i {
            break
        }
        // Hoán đổi hai nút
        nums.swapAt(i, ma)
        // Tiếp tục vun đống xuống dưới
        i = ma
    }
}

/* Sắp xếp vun đống (heap sort) */
func heapSort(nums: inout [Int]) {
    // Thao tác xây dựng đống: vun đống tất cả các nút trừ các nút lá
    for i in stride(from: nums.count / 2 - 1, through: 0, by: -1) {
        siftDown(nums: &nums, n: nums.count, i: i)
    }
    // Lấy ra phần tử lớn nhất từ đống, lặp lại n-1 vòng
    for i in nums.indices.dropFirst().reversed() {
        // Xóa nút
        nums.swapAt(0, i)
        // Bắt đầu vun đống nút gốc, từ trên xuống dưới
        siftDown(nums: &nums, n: i, i: 0)
    }
}

@main
enum HeapSort {
    /* Mã điều khiển */
    static func main() {
        var nums = [4, 1, 3, 1, 5, 2]
        heapSort(nums: &nums)
        print("Sau khi sắp xếp vun đống, nums = \(nums)")
    }
}

File: vi/codes/cpp/chapter_sorting/heap_sort.cpp
/**
 * File: heap_sort.cpp
 * Created Time: 2023-05-26
 * Author: krahets (krahets@163.com)
 */

#include "../utils/common.hpp"

/* Độ dài heap là n, bắt đầu heap hóa nút i, từ trên xuống dưới */
void siftDown(vector<int> &nums, int n, int i) {
    while (true) {
        // Nếu nút i là lớn nhất hoặc chỉ số l, r vượt quá giới hạn, không cần tiếp tục heap hóa, thoát vòng lặp
        int l = 2 * i + 1;
        int r = 2 * i + 2;
        int ma = i;
        if (l < n && nums[l] > nums[ma])
            ma = l;
        if (r < n && nums[r] > nums[ma])
            ma = r;
        // Hoán đổi hai nút
        if (ma == i) {
            break;
        }
        // Hoán đổi hai nút
        swap(nums[i], nums[ma]);
        // Vòng lặp heap hóa xuống dưới
        i = ma;
    }
}

/* Sắp xếp vun đống (heap sort) */
void heapSort(vector<int> &nums) {
    // Thao tác xây dựng heap: heap hóa tất cả các nút trừ nút lá
    for (int i = nums.size() / 2 - 1; i >= 0; --i) {
        siftDown(nums, nums.size(), i);
    }
    // Trích xuất phần tử lớn nhất trong heap và lặp lại n-1 vòng
    for (int i = nums.size() - 1; i > 0; --i) {
        // Xóa nút
        swap(nums[0], nums[i]);
        // Bắt đầu heap hóa nút gốc, từ trên xuống dưới
        siftDown(nums, i, 0);
    }
}

/* Mã chạy chương trình */
int main() {
    vector<int> nums = {4, 1, 3, 1, 5, 2};
    heapSort(nums);
    cout << "After heap sort completes, nums = ";
    printVector(nums);

    return 0;
}
